modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/PredicateExpressionVisitorImpl.java

```diff
@@ -458,59 +473,65 @@ private Predicate _getInPredicate(
 		);
 	}
 
-	private EntityModel _getObjectDefinitionEntityModel(
-		long objectDefinitionId) {
+	private List<ObjectValuePair<ObjectRelationship, Long>>
+		_getListObjectRelationshipsRelatedObjectDefinitionId(
+			long objectDefinitionId, List<String> objectRelationshipsNames) {
 
-		EntityModel entityModel = _entityModels.get(objectDefinitionId);
+		List<ObjectValuePair<ObjectRelationship, Long>> objectValuePairs =
+			new ArrayList<>();
 
-		if (entityModel == null) {
-			entityModel = _createEntityModel(objectDefinitionId);
+		for (String objectRelationshipsName : objectRelationshipsNames) {
+			ObjectRelationship objectRelationship = _fetchObjectRelationship(
+				objectDefinitionId, objectRelationshipsName);
 
-			_entityModels.put(objectDefinitionId, entityModel);
-		}
+			if (objectRelationship != null) {
+				objectValuePairs.add(
+					new ObjectValuePair<>(
+						objectRelationship, objectDefinitionId));
 
-		return entityModel;
-	}
+				objectDefinitionId = _getRelatedObjectDefinitionId(
+					objectDefinitionId, objectRelationship);
+			}
+		}
 
-	private Predicate _getObjectRelationshipPredicate(
-		Object left,
-		UnsafeBiFunction<String, Long, Predicate, Exception> unsafeBiFunction) {
+		if (objectValuePairs.isEmpty()) {
+			throw new ServerErrorException(
+				500,
+				new Exception(
+					"Unexpected error. ObjectRelationships not founded"));
+		}
 
-		String leftString = (String)left;
+		Collections.reverse(objectValuePairs);
 
-		String[] leftStringParts = leftString.split(StringPool.SLASH);
+		return objectValuePairs;
+	}
 
-		String relationshipName = leftStringParts[0];
+	private EntityModel _getObjectDefinitionEntityModel(
+		long objectDefinitionId) {
 
-		ObjectRelationship objectRelationship = _fetchObjectRelationship(
-			relationshipName);
+		EntityModel entityModel = _entityModels.get(objectDefinitionId);
 
-		if (objectRelationship != null) {
-			String objectFieldName = leftStringParts[1];
+		if (entityModel == null) {
+			entityModel = _createEntityModel(objectDefinitionId);
 
-			try {
-				return _getObjectRelationshipPredicate(
-					objectRelationship,
-					unsafeBiFunction.apply(
-						objectFieldName,
-						_getRelatedObjectDefinitionId(
-							_objectDefinitionId, objectRelationship)));
-			}
-			catch (Exception exception) {
-				throw new RuntimeException(exception);
-			}
+			_entityModels.put(objectDefinitionId, entityModel);
 		}
 
-		return null;
+		return entityModel;
 	}
 
 	private Predicate _getObjectRelationshipPredicate(
+			long objectDefinitionId,
+			List<ObjectValuePair<ObjectRelationship, Long>>
+				objectRelationshipsObjectsDefinitionIdsList,
 			ObjectRelationship objectRelationship, Predicate predicate)
 		throws Exception {
 
+		objectRelationshipsObjectsDefinitionIdsList.remove(0);
```

Comment:
Maybe it is better to call remove from outside as it already returns the element

Reply:
Done! 🙂